### PR TITLE
App Check: App Attest provider

### PR DESCRIPTION
**Pull request by maksymmalyhin**
- support of [Apple's App Attest](https://developer.apple.com/documentation/devicecheck/preparing_to_use_the_app_attest_service)  attestation provider.

### CODE REVIEW

**Review by ncooke3 — APPROVED** [review]
**LGTM** I've tagged any TODO's that were commented in case we'd like to resolve them/set up tasks to track them.  No further action needed to merge this in– just want to make sure we don't forget anything!

**Comment by maksymmalyhin** [comment]
@ncooke3 Thank you for the review and spotting the unattended TODOs. I'll add another PR into `appcheck-appattest-main` with the cleanup.

File: FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+#import "FirebaseAppCheck/Sources/Core/Errors/FIRAppCheckErrorUtil.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+// TODO: Verify the following request fields.
+static NSString *const kRequestFieldArtifact = @"artifact";
+static NSString *const kRequestFieldAssertion = @"assertion";
+static NSString *const kRequestFieldAttestation = @"attestation_statement";
+static NSString *const kRequestFieldChallenge = @"challenge";
+static NSString *const kRequestFieldKeyID = @"key_id";
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m:38]: Actually no, it looks like the backend accepts both variants.

File: FirebaseAppCheck/Apps/FIRAppCheckTestApp/Podfile [hunk 2]
```diff
@@ -11,7 +11,5 @@ target 'FIRAppCheckTestApp' do
   pod 'FirebaseAppCheck', :path => '../../../'
 
   pod 'FirebaseCore', :path => '../../../'
```
Comment (paulb777) [inline FirebaseAppCheck/Apps/FIRAppCheckTestApp/Podfile:13]: Not for this PR, but there should be a build test in CI to keep this unstale.

File: FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRAppAttestProvider.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
+/*
```
Comment (paulb777) [inline FirebaseAppCheck/Sources/Public/FirebaseAppCheck/FIRAppAttestProvider.h:1]: Please add to the umbrella header.

**Review by paulb777 — APPROVED** [review]
LGTM after umbrella header is updated.

File: FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  }
+  return HTTPBodyPromise;
+}
+
+#pragma mark - Helpers
+
+- (NSString *)base64StringWithData:(NSData *)data {
+  // TODO: Need to encode in base64URL?
+  return [data base64EncodedStringWithOptions:0];
+}
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/AppAttestProvider/API/FIRAppAttestAPIService.m]: Addressed in #8170.

File: FirebaseAppCheck/Sources/AppAttestProvider/FIRAppAttestProvider.m [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+                         ]]
+      .thenOn(self.queue,
+              ^FBLPromise<FIRAppAttestKeyAttestationResult *> *(NSArray *challengeAndKeyID) {
+                // 2. Attest the key.
+                NSData *challenge = challengeAndKeyID.firstObject;
+                NSString *keyID = challengeAndKeyID.lastObject;
+
+                return [self attestKey:keyID challenge:challenge];
+              })
+      // TODO: Handle a possible key rejection - generate another key.
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply FirebaseAppCheck/Sources/AppAttestProvider/FIRAppAttestProvider.m]: Addressed in #8170.

File: FirebaseAppCheck/Tests/Unit/Core/FIRAppCheckTests.m [hunk 6]
```diff
@@ -358,18 +335,17 @@ - (void)testGetToken_AppCheckProviderError {
...
-                    XCTAssertNotNil(result);
-                    XCTAssertEqualObjects(result.token, @"eyJlcnJvciI6IlVOS05PV05fRVJST1IifQ==");
+                               XCTAssertNotNil(result);
+                               XCTAssertEqualObjects(result.token, kDummyToken);
 
-                    // TODO: Expect a public domain error to be returned - not the
-                    // internal one.
-                    XCTAssertEqualObjects(result.error, providerError);
-                  }];
+                               // TODO: Expect a public domain error to be returned - not the
```
[the remark this answers is not in the document]
Comment (maksymmalyhin) [reply FirebaseAppCheck/Tests/Unit/Core/FIRAppCheckTests.m]: It is still relevant. Currently we don't have public API for this method. The TODO is expected to be addressed when we introduce the public API.

**Review by ncooke3 — APPROVED** [review]
LGTM on green

**Pull request merged**